### PR TITLE
ARROW-13876: [C++] Add trivial null kernels to arithmetic, sort functions

### PR DESCRIPTION
**Pull request by lidavidm**

### CODE REVIEW

**Comment by westonpace** [comment]
This is `ALL null -> null` correct?  I.e. this adds support for `null + null => null` but it does not add support for `null + 7 => null`?  Do we support casting to null in dispatch best?

**Comment by lidavidm** [comment]
Yes, this is the NA/Null type. For mixed nulls, that sort of thing should generally be supported already, at least for binary functions.

**Comment by westonpace** [comment]
Right.  But for addition you could have...

```
arr1 = pa.array([None, None], pa.null
arr2 = pa.array([1, 2])
pc.add(arr1, arr2)
```

I don't think this covers that, correct?

**Comment by lidavidm** [comment]
This doesn't cover that because it's already supported, even in 5.0:

```
>>> import pyarrow as pa
>>> import pyarrow.compute as pc
>>> arr1 = pa.array([None, None], pa.null())
>>> arr2 = pa.array([1, 2])
>>> pc.add(arr1, arr2)
<pyarrow.lib.Int64Array object at 0x7f1e632df100>
[
  null,
  null
]
```

**Comment by lidavidm** [comment]
Though if this isn't properly supported in your testing then we should probably take care of that.

**Comment by westonpace** [comment]
My type testing typically doesn't test mixed types (e.g. `add(int64, int32)`) any longer (there are some exceptions) since I assume that is handled mostly by casting.  If this is getting handled because dispatch best is casting null up to int64 (and is willing to cast null up to any data type) then I don't think I need to test it.  On the other hand, if this is handled by registering some explicit kernels to handle "some null" then I probably should include it.

**Comment by lidavidm** [comment]
Ah, ok. Those cases are indeed handled by casting in DispatchBest, not by registering individual kernels. This PR registers a separate kernel for the all-null-type case mostly because it's easiest and there isn't really an obvious type to cast to otherwise.

File: cpp/src/arrow/compute/kernels/vector_sort_test.cc [hunk 1]
```diff
@@ -675,6 +687,15 @@ TYPED_TEST(TestArraySortIndicesForDecimal, DecimalSortTestTypes) {
...
 
+TEST(TestArraySortIndices, Null) {
+  for (const auto null_placement : AllNullPlacements()) {
+    for (const auto order : AllOrders()) {
+      AssertSortIndices(null(), "[null, null, null, null]", order, null_placement,
+                        "[0, 1, 2, 3]");
+    }
+  }
+}
+
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/vector_sort_test.cc:698]: Can you also add tests for chunked array, record batch and table sorting?

Comment (lidavidm) [reply cpp/src/arrow/compute/kernels/vector_sort_test.cc:759]: Ah, good point. I've added the necessary support for those cases.

**Review by pitrou — APPROVED** [review]

**Pull request closed without merging**